handler/handler.go

```diff
@@ -151,33 +148,111 @@ func (h *InstanceComplete) ValidateInstance(i dataset.Instance) error {
+// ParseQueryResponse parses the provided cantabular response into a CSV bufio.Reader,
+// where the first row corresponds to the dimension names header (including a count)
+// and each subsequent row corresponds to a unique combination of dimension values and their count.
+//
+// Example by Sensible Code here: https://github.com/cantabular/examples/blob/master/golang/main.go
+func (h *InstanceComplete) ParseQueryResponse(resp *cantabular.StaticDatasetQuery) (*bufio.Reader, error) {
+	// Create CSV writer with underlying buffer
+	b := new(bytes.Buffer)
+	w := csv.NewWriter(b)
+
+	// Obtain the CSV header
+	header := createCsvHeader(resp.Dataset.Table.Dimensions)
+	w.Write(header)
+	if err := w.Error(); err != nil {
+		return nil, fmt.Errorf("error writing the csv header: %w", err)
+	}
+
+	// Obtain the CSV rows according to the cantabular dimensions and counts
+	for i, count := range resp.Dataset.Table.Values {
+		row := createCsvRow(resp.Dataset.Table.Dimensions, i, count)
+		w.Write(row)
+		if err := w.Error(); err != nil {
+			return nil, fmt.Errorf("error writing a csv row: %w", err)
+		}
+	}
+
+	// Flush to make sure all data is present in the byte buffer
+	w.Flush()
+	if err := w.Error(); err != nil {
+		return nil, fmt.Errorf("error flushing the csv writer: %w", err)
+	}
 
-	// Using a bufio.ReadWriter we should be able to use an object that can be written
-	// to by the csv package and then uploaded directly with the S3 package.
-	var csv bufio.ReadWriter
-	return csv, nil
+	// Return a reader with the same underlying Byte buffer as written by the csv writter
+	return bufio.NewReader(b), nil
+}
+
+// createCsvHeader creates an array of strings corresponding to a csv header
+// where each column contains the value of the corresponding dimension, with the last column being the 'count'
+func createCsvHeader(dims []cantabular.Dimension) []string {
+	header := make([]string, len(dims)+1)
+	for i, dim := range dims {
+		header[i] = dim.Variable.Label
+	}
+	header[len(dims)] = "count"
+	return header
+}
+
+// createCsvRow creates an array of strings corresponding to a csv row
+// for the provided array of dimension, index and count
+// it assumes that the values are soreted with lower weight for the last dimension and higher weight for the first dimension.
+func createCsvRow(dims []cantabular.Dimension, index, count int) []string {
+	row := make([]string, len(dims)+1)
+	for i := len(dims) - 1; i >= 0; i-- {
+		row[i] = dims[i].Categories[index%dims[i].Count].Label
+		index = index / dims[i].Count
+	}
+	row[len(dims)] = fmt.Sprintf("%d", count)
+	return row
 }
 
 // UploadCSVFile uploads the provided file content to AWS S3
 // The file name is the instance ID and a uuid
-func (h *InstanceComplete) UploadCSVFile(ctx context.Context, instanceID string, file *bufio.ReadWriter, encrypted bool) (string, error) {
+func (h *InstanceComplete) UploadCSVFile(ctx context.Context, instanceID string, fReader io.Reader, encrypted bool) (string, error) {
```

Comment:
Does this really need to be called fReader? I think file is still probably a more appropriate name

Reply:
I think you're right - I tried to make it clear it was a reader, but the type already defines it. I will change it back